### DISCUSSION TITLE
Automatically run `session.commit()` in session dependency BEFORE returning request

### DISCUSSION THREAD

**Discussion by anguspmitchell** [fastapi/fastapi#8766] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# orm.py
def fastapi_session_scope() -> sqlalchemy.orm.session.Session:
    session = session_maker(bind=engine)
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()

# routers.py
@router.post("/v1/organizations", response_model=Organization)
def create_org(session: Session = Depends(orm.fastapi_session_scope)):
    org = sqlalc.Organization(name="test")
    session.add(org)
    return org

@router.get("/v1/organizations", response_model=List[Organization])
def get_orgs(session: Session = Depends(orm.fastapi_session_scope)):
    return session.query(sqlalc.Organization).all()
```

### Description

(I realize this is diverging slightly from the [SQLAlchemy example](https://fastapi.tiangolo.com/tutorial/sql-databases/), which explicitly uses `session.commit()` in the body of the function and not in the session dependency.)

I'm creating a row with a POST and then immediately afterwards fetching all rows with a GET. I've got `session.commit()` in the session scope dependency, which is convenient because it runs a `commit()` at the end no matter what and doesn't require us to always explicitly commit at the end of an endpoint body.

As written, if an Organization is created via a POST and then all the organizations are immediately afterwards fetched via a GET, the GET will not include the recently created Organization. This is not a JS async issue. What's happening, as best I can tell, is that the POST starts a transaction, `org` is returned, and then, before the `commit()` actually happens, the GET is called and the list of organizations is returned without the organization that was just created (or so it seemed). 

```
t  POST            GET

1  request
2  session.add
3  return
4                  request
5                  session.query
6                  return          # POSTed organization is missing!
7  session.commit
8                  session.commit
```

The behavior I'd like is for `commit()` to run BEFORE the value is returned, which would guarantee that the POST and GET could be run back-to-back safely. Is that possible using a FastAPI dependency? Has anyone else run into this issue?

```
t  POST            GET

1  request
2  session.add
3  session.commit
4  return
5                  request
6                  session.query
7                  session.commit
8                  return

```

### Operating System

macOS

### FastAPI Version

0.59.0

### Python Version

3.8.10

**Comment by waynerv** [comment, marked as answer]
As the [documentation](https://fastapi.tiangolo.com/zh/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-httpexception) says, the statement after yield will be executed after the response is returned, so you should consider putting the commit statement in the function body

**Comment by anguspmitchell** [comment]
@waynerv Got it, thanks. It sounds like the options are either 1) to use `Depends()` and always specify `commit()` in the endpoint body or 2) to use a context manager, an explicit with statement and an implicit `commit()` (like the following)

```python
# orm.py
@contextmanager
def session_scope(expire_on_commit: bool = True) -> sqlalchemy.orm.session.Session:
    orm_session = _orm_session(expire_on_commit=expire_on_commit)
    try:
        yield orm_session
        orm_session.commit()
    except:
        orm_session.rollback()
        raise
    finally:
        orm_session.close()

# routers.py
@router.post("/v1/organizations", response_model=Organization)
def create_org():
    with orm.session_scope() as session:
        org = sqlalc.Organization(name="test")
        session.add(org)
        return Organization.from_orm(org)
```

**Comment by AlexanderPodorov** [comment]
@anguspmitchell , I think you should avoid dependencies like this:
```python
@contextmanager
def session_scope(expire_on_commit: bool = True) -> sqlalchemy.orm.session.Session:
    orm_session = _orm_session(expire_on_commit=expire_on_commit)
    try:
        yield orm_session
        orm_session.commit()
    except:
        orm_session.rollback()
        raise
    finally:
        orm_session.close()
```
For those reasons:
1. as @waynerv said, the statement after yield will be executed after the response is returned. So that means, in some circumstances you will send OK response, but commit will actually fail.
2. If you raise `HTTPException` from endpoint function body, then session **will be commited too**. This kind of exception won't be caught by `try-except` of `session_scope` function.

P.S. I also tried to use the same approach in the very beginning of my work with FastAPI, but finally I'm confident it's much better to commit explicitly at the end of endpoint function body.

**Comment by anguspmitchell** [comment]
@AlexanderPodorov Very good points.

I may not have been entirely clear on this part:

> It sounds like the options are either 1) to use Depends() and always specify commit() in the endpoint body or 2) to use a context manager, an explicit with statement and an implicit commit()

What I'm proposing in (2) is just to not use a FastAPI dependency at all, and instead to use an explicit `with` block in the function body. (More verbose obviously, but, alas, tradeoffs.) I really do not want to run into the problems that you mentioned, though. To make sure those issues don't occur with this approach, I added two lines of code to test it. From what I can tell, neither of dangers you mentioned come up when using an explicit `with` block:

```python
# orm.py
@contextmanager
def session_scope(expire_on_commit: bool = True) -> sqlalchemy.orm.session.Session:
    orm_session = _orm_session(expire_on_commit=expire_on_commit)
    try:
        yield orm_session
        raise   # <- this causes a 500 error to be returned, 
                #    so it doesn't appear to run the risk of returning OK
                #    but then silently failing after the yield statement
        orm_session.commit()
    except:
        orm_session.rollback()
        raise
    finally:
        orm_session.close()

# routers.py
@router.post("/v1/organizations", response_model=Organization)
def create_org():
    with orm.session_scope() as session:
        org = sqlalc.Organization(name="test")
        session.add(org)
        raise HTTPException(status_code=404, detail="Item not found") # <- this returns a 404
                    # and causes the rollback in the context manager to be executed, 
                    # avoiding the possibility of a commit after an HTTPException raise
        return Organization.from_orm(org)
```

**Comment by AlexanderPodorov** [comment]
Yes, I think this approach should work. And yes, you can create a session in the endpoint function. My 1. and 2. points are related to dependency approach.
BTW, as of SQLAlchemy 1.4 we don't need to write our own context managers for sessions/transactions, it should work out of box.
See: https://docs.sqlalchemy.org/en/14/orm/session_transaction.html
Example for their site:
```python
Session = sesssionmaker(engine)

with Session() as session:
    with session.begin():
        session.add(some_object)
    # commits

# closes the Session
```

**Comment by arsenron** [comment]
You can add database session to ```starlette.Request.state``` in ```get_db``` dependency and wrap all your requests with a middleware that checks if ```starlette.Request.state``` contains your database session and if so, commit it or whatever you want. I wrote for an async database but it can be applied to sync. For example:
```python
import uvicorn
from fastapi import FastAPI, Request, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import sessionmaker

from database import engine  # some engine

app = FastAPI()

@app.middleware("http")
async def commit_db_session(request: Request, call_next):
    """
    Commit db_session before response is returned.
    """
    response = await call_next(request)
    db_session = request.state._state.get("db_session")  # noqa
    if db_session:
        await db_session.commit()
    return response


async_session = sessionmaker(
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=False,
)


async def get_db(request: Request) -> AsyncSession:
    session: AsyncSession = async_session()
    # add session to starlette.Request.state
    request.state.db_session = session
    yield session


@app.get("/")
async def check_db(db: AsyncSession = Depends(get_db)):
    return {"message": "db session is closed automatically before returning response"}


if __name__ == "__main__":
    uvicorn.run(app)
```

**Reply by CannonLock** [reply]
This did the trick! Thanks.

**Comment by skion** [comment]
I am aware this behaviour is documented, but not being able to raise exceptions after yielding in dependencies does seem like a fantastic gun to shoot yourself in the foot with. The risk is of course that the API (already) returns a positive response to the user, while the data is not yet written to the database, which might go unnoticed for a long time and lead to data sporadically disappearing on well-timed connection issues for instance.

The current behaviour also seems to be at friction with [SQLAlchemy's documentation on session lifespan management](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it).

Besides being unexpected (at least for some), there does not appear to be an obvious workaround. The workaround that @arsenron proposes is of course excellent (and I'm using that now), but also very asymmetric and brittle because of the logic-split.

The alternative of placing explicit `session.commit()`'s in your views is also risky in my opinion, since if you forget one then you risk losing already acknowledged data once again.

I believe that this is exactly where a framework should prevent the developer from doing something silly. The whole point of using yield in a dependency is so that you can do things after it, committing data (and handling its exceptions properly) being one of those things.

As an anecdotal proof-point of how unexpected this is, a quick search indicates that for example [FastAPI-Restful appears to be at risk of data-loss](https://github.com/yuval9313/FastApi-RESTful/blob/19655d19c78d95de15e45b71c002e3324c87a099/fastapi_restful/session.py#L143) because of this issue since they commit-after-yield as well.

I hope we can improve this situation somehow, perhaps by making background processing optional? Just thinking out loud here, there are probably better options.

**Comment by Lawouach** [comment]
Hi all,

I was made aware of that behavior through a bit of tracing when I realised my db sessions were committed after the response was returned. One of the other downside is that we keep a transaction opened longer than it needs when it is would be better to release db resources as early as possible. i'll go with the proposed workaround as well but I believe this is a bit of an issue that should at least be documented (I mean how the documented behavior may lead to potential risks).

**Comment by skion** [comment]
FWIW I resorted to using a middleware in the end, and picking up the session in a dependency through the application state:

```python
@app.middleware("http")
async def db_session(request: Request, call_next):

    with _SessionLocal() as session, session.begin():

        # Make the session available to the FastAPI dependency.
        request.state.db_session = session

        return await call_next(request)


def get_database_session(request: Request) -> Session:
    """
    FastAPI dependency to obtain a database connection.
    """
    # This is set up in the database middleware.
    return request.state.db_session
```

This felt simpler than the solution proposed before, and keeps the session logic in one place so you can use the context managers again. Caveat is that database access in background tasks is probably completely broken without further work.

**Reply by shaiArn** [reply]
Hello @skion ,

I came across your post from January about handling SQLAlchemy sessions in FastAPI using middleware. The method you described seems simple and elegant, and I was considering implementing something similar. I wanted to follow up and see if you're still using this solution six months later? If so, how has it been holding up in production? Have you encountered any issues, or made any significant updates or improvements? Any feedback would be greatly appreciated!

**Reply by skion** [reply]
@shaiArn We're still using the above method indeed, also in production and so far without any issues. That said we have not performed any real stress tests either...

**Reply by aeris** [reply]
Hello here!

I faced this trouble today and yes, it leads to data-loss with no real possible easy workaround.
I can't use the middleware way because my database connection depends too of route variable or another dependency, something like 

```
def get_session(id: str) -> Session:
    session_maker = get_session_maker_from_id(id)
    with session_maker() as session:
        yield session
        session.commit()


@router.get("/{id}")
def hello(id: str, session = Depends(get_session):
    ...
```

or worst

```
def get_user(request: Request) -> User:
    ...

def get_session(user = Depends(get_user)) -> Session:
    session_maker = get_session_maker_from_user(user)
    with session_maker() as session:
        yield session
        session.commit()


@router.get()
def hello(user = Depends(get_user), session = Depends(get_session)):
    ...
```

Because of this yield/exception trouble, I hit a SQL error (missing default SQL value on insert, not detected before the commit) on the commit and the overall session was rollbacked, but the response was 200 from the client point of view with all the usual information. Later, he fetched the supposed created resource and just faced a 404…
Hopefully it happen in test environment, if this happen in production, it would be probably catastrophic.

I manage to workaround this with a context manager proxy around the `session_maker` to handle the `with session: commit` outside the depends, but it's very tedious to generalize that for each context manager-like dependencies and it's very error prone, having to remember of the real behavior each time you `try/except` in a dependency is not very safe.

Nobody sane expect such behavior even it is documented (but lost in the middle of a page with I guess nobody going at this point of the page, given all seems working just after the 2nd paragraph with explicit `try/except` without a single warning) and IMO there is no real decent reason to trigger background jobs before a full dependency execution (for example to be able to modify/wrap response after the yield).

**Reply by aeris** [reply]
Correction: even with a context manager wrapper, the exception is still after the yield and so just ignored… :expressionless:

**Reply by AdityaSoni19031997** [reply]
Thanks for the discussions above. I'm curious, If we're to do this in 2025, have things changed for better or the above is still the best way to go about?

Thanks

**Reply by aeris** [reply]
It was fixed at some point on Fastapi, and you haven't to bother about this now.
Will try to find the commit.

**Reply by aeris** [reply]
If I remember correctly, it's this one : https://github.com/fastapi/fastapi/pull/10831/commits/6524d0fe7a2348e61d1ed3d56233dd490a7f8795

https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#background-tasks-and-dependencies-with-yield-technical-details

**Reply by AdityaSoni19031997** [reply]
Thanks a lot for sharing an update!

**Comment by yinziyan1206** [comment]
I'd like to add service layer and then i can create session in one service.
like this:
```python
class ServiceMetaClass(ABCMeta):
    __bind_key__ = "default"
    __root__ = None

    def __new__(mcs, name: str, bases: tuple[type, ...], attrs: dict, **kwargs):
        new_cls = super().__new__(mcs, name, bases, attrs)
        if ds := kwargs.get("datasource"):
            new_cls.__bind_key__ = ds
        return new_cls

    def __call__(cls, *args, **kwargs):
        if cls.__root__ is None:
            cls.__root__ = super().__call__(*args, **kwargs)
        return cls.__root__

    @property
    def session(cls) -> AsyncContextManager[AsyncSession]:
        return execute_wrapper(cls.__bind_key__)

class BaseService(metaclass=ServiceMetaClass):
    ...
```

**Comment by Danstiv** [comment]
Middleware solution is bad, because it hides exceptions for some reason, and we are committing transaction, for example, when HTTPException is raised from dependency.

**Comment by skion** [comment]
FWIW, this topic is apparently either back or was never really gone, and is actively being discussed [here](https://github.com/fastapi/fastapi/discussions/14151).
